FLUID-5525: add OpenDyslexic as an option to text-font panel

waharnum opened the pull request:
Add the OpenDyslexic font as an option to the preferences framework text-font panel, based on work to add it to CISL's version.

Sources OpenDyslexic from the NPM package, pinned to latest version.

Resolves (adding, rather than replacing) this JIRA issue: https://issues.fluidproject.org/browse/FLUID-5525

waharnum commented:
@jobara, can you review this for merge?

jobara commented:
@waharnum could you also update the release with mention of the license and which version of Open Dyslexic we are using. The 3rd party software is all organized by license in there.

waharnum commented:
There seems to be an ongoing problem with the CI failing that isn't necessarily related to this PR. I am investigating now.

waharnum commented:
ok to test

waharnum commented:
CI tests pass when Headless Chrome is excluded as a browser to use and consistently fail when it is present (hangs forever). Appears to occur on the `SyllabificationEnactor-test` run.

Have opened separate PR to address what I think is an underlying issue in the testem config: https://github.com/fluid-project/infusion/pull/957

jobara commented on src/framework/preferences/js/PrefsEditor.js:
```diff
@@ -441,6 +441,7 @@ var fluid_3_0_0 = fluid_3_0_0 || {};
 
     // called once markup is applied to the document containing tab component roots
     fluid.prefs.prefsEditor.finishInit = function (that) {
+        console.log(that);
```
stray console.log

jobara commented:
@waharnum with the OpenDyslexic font being pulled in via NPM we shouldn't need to commit the font files to git. Could you please update any ignore (e.g. .gitignore) as needed. Also we need to makes sure that the fonts are properly added to the dist directory. Perhaps you could also address https://issues.fluidproject.org/browse/FLUID-6361 at the same time. A more general fix for these will be done in https://issues.fluidproject.org/browse/FLUID-6359 but if you can get fonts fixed up ahead of time that would be helpful.

jobara commented:
@waharnum seems this last build failure was related to linting

waharnum commented:
@jobara I've done the following re the `dist` directory:

- added OpenDyslexic and OpenSans to what is copied over (so this PR should also address https://issues.fluidproject.org/browse/FLUID-6361 now)
- added tasks to grunt to verify the fonts are present in `dist`, along the lines of the existing JS and CSS verification tasks
- refactored the grunt tasks for file verification to eliminate logic duplication

jobara commented on Gruntfile.js:
```diff
@@ -166,7 +167,7 @@ module.exports = function (grunt) {
                     expand: true,
                     cwd: "build/",
-                    src: ["src/lib/fonts/**", "src/framework/preferences/fonts/**", "src/framework/preferences/images/**"],
+                    src: ["src/lib/fonts/**", "src/framework/preferences/fonts/**", "src/framework/preferences/images/**", "src/lib/open-dyslexic/**", "src/lib/opensans/**"],
```
I don't think the "src/lib/fonts" directory exists anymore.

jobara commented:
@waharnum I noticed that the full page instances are having problems. The previews do not display in openDyslexic and the minimal footprint content page doesn't either. 

- infusion/tests/manual-tests/framework/preferences/fullPage/
- infusion/tests/manual-tests/framework/preferences/fullPage-schema/
- infusion/examples/framework/preferences/minimalFootprint/

waharnum commented:
@jobara This should now be fixed - the `font-face` definitions should have been in the `Enactors.stylus` file rather than the `PrefsEditor.stylus` file - OpenDyslexic was working on the demo page because that includes `PrefsEditor.css` in the page, but those other contexts don't.

jobara commented on src/framework/preferences/css/stylus/Enactors.styl:
```diff
@@ -32,3 +32,32 @@ build-themes-Enactors(themes);
+    src:url('../../../lib/open-dyslexic/fonts/OpenDyslexic-BoldItalic.woff');
+    font-weight: 600;
+    font-style: italic;
+}
```
I wonder if this font-face definition should be part of build-fonts-Enactors method, or at least somehow part of Fonts.styl. It seems that if one were to use that stylus helper but not have manually added the font definition, that it wouldn't work properly.

waharnum replied:
This seems reasonable. It could probably set up as a generator-style function as well (though OpenDyslexic would be the only example).

jobara replied:
I believe this PR is just waiting on addressing this comment.

jobara commented on Gruntfile.js:
```diff
@@ -652,17 +669,36 @@ module.exports = function (grunt) {
+                message: "Verify opensans font files are present in /dist/ directory",
+                distDirectory: "dist/assets/src/lib/opensans/fonts"
+            }
+        };
```
Looking at this, seems a bit odd that the configuration for which files to check isn't in the typical grunt config. I think ideally that would be the case for all the verification steps. Is there a reason why the configuration is hardcoded in each task?

waharnum replied:
Is your suggestion simply to move the functions etc that generate the file lists into the Grunt config, or to move to manually maintaining the list of files to verify? (I'm unclear as to which)

jobara replied:
I mean would it be possible to make use of grunts own [files](https://gruntjs.com/configuring-tasks#files) mechanics.

waharnum replied:
I've finished initial refactoring to use Grunt's file mechanics where possible. Where not (the generated JS files), I've made the multitask accept a custom `verifyFilesListFunc`. 

So task configuration either uses the standard Grunt file mechanics where there's a one-to-one mapping (possibly with renaming) between source files and destination files, or specifies a function to return an array of files to check. I think this should have the flexibility required for future additions - most of them will fall into the first case, but the second is available if the verification logic is not as straightforward as "these source files are compiled into these destination files".

Let me know your thoughts.

jobara commented on Gruntfile.js:
```diff
@@ -469,6 +475,48 @@ module.exports = function (grunt) {
+                        expand: true,
+                        src: ["*.styl"],
+                        cwd: "src/framework/preferences/css/stylus/",
+                        dest: "dist/assets/src/framework/preferences/css/stylus/",
```
There is also a stylus file from the Switch component that gets compiled and placed in the Assets directory. We should probably test for this one too.

jobara commented on Gruntfile.js:
```diff
@@ -618,46 +663,48 @@ module.exports = function (grunt) {
+        if (this.files.length > 1) {
+            _.forEach(this.files, function (file) {
+                expectedFiles.push(file.dest);
+            });
```
Minor change, but I think you can just use either [_.map](https://lodash.com/docs/4.17.15#map) or JavaScript's array [map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) method.

jobara commented on Gruntfile.js:
```diff
@@ -469,6 +475,48 @@ module.exports = function (grunt) {
+                        var distDirectory = "dist/";
+                        expectedFilenames.push(distDirectory + jsFilename, distDirectory + mapFilename);
+                    });
+                    return expectedFilenames;
```
Similar to previous comment, can use the map method instead of forEach.

jobara replied:
May want to consider defining this function outside of the config and just passing in the reference to it here.

jobara commented:
@waharnum could you please merge master, and maybe file a JIRA about moving the file verification into a grunt module?

waharnum commented:
Have created Grunt module JIRA as https://issues.fluidproject.org/browse/FLUID-6387

jobara commented on src/framework/preferences/css/stylus/utils/Fonts.styl:
```diff
@@ -27,5 +60,21 @@ build-fonts-Enactors(textFont) {
+    }    
+
+    // Build any necessary @font-face definitions
+    for fontFamilyName in textFontFaces {
```
Things seem to be working, so I'm not sure if this really matters, but as set up, the `@font-face` declaration comes after using the fonts in the CSS file.

waharnum replied:
It's probably more intuitive to have it come before - I'll make the change.

jobara commented:
@waharnum for the list verification we don't seem to be testing the minified css files nor the PrefsFramework-Icons.woff file.

jobara commented:
@waharnum when the OpenDyslexic font is applied to the Prefs Editor panel, the contrast selection styling is off. The options look like lego heads, which is kind of cool, but maybe not what we want.
![contrast](https://user-images.githubusercontent.com/556024/62050886-2d299000-b1e0-11e9-8f23-f04d9eaeba31.png)

jobara commented:
In Firefox the text style options are displayed in their own font, except for OpenDyslexic. I'm not sure if this is a limitation of the browser only showing the system fonts, or if there is something not setup correctly. 
![textStyleOptions](https://user-images.githubusercontent.com/556024/62050940-4fbba900-b1e0-11e9-8445-546a2fb9a621.png)

In MS Edge, it shows up fine in the original list, but after applying the OpenDyslexic font, the whole lists shows up in OpenDyslexic instead of displaying in their own font types. Again, could be related to OpenDyslexic not being installed on the system.
![MS_Edge_textStyle](https://user-images.githubusercontent.com/556024/62051101-a45f2400-b1e0-11e9-82a3-e31c5c23988a.png)

jobara commented:
@waharnum maybe just file that in a separate JIRA then, seems suitably different from this work if it's already an issue.

waharnum commented:
> 
> 
> @waharnum maybe just file that in a separate JIRA then, seems suitably different from this work if it's already an issue.

Filed as https://issues.fluidproject.org/browse/FLUID-6388

waharnum commented:
> 
> 
> @waharnum for the list verification we don't seem to be testing the minified css files nor the PrefsFramework-Icons.woff file.

I've added the minified CSS files and the icon font file to the verification config.

waharnum commented:
> @waharnum when the OpenDyslexic font is applied to the Prefs Editor panel, the contrast selection styling is off. The options look like lego heads, which is kind of cool, but maybe not what we want.

Open Dyslexic seems to have unusually tall character blocks, which were extending outside the container. Adding `overflow:hidden` to their style seems to have fixed the issue.

jobara commented:
merged at 058483bcca139731896ad11e4f734034e1d5cc0b
